tracing.go

```diff
@@ -84,53 +87,155 @@ func (s Span) ParentID() uint64 { return s.parentID }
 // SpanID returns the id of this specific RPC
 func (s Span) SpanID() uint64 { return s.spanID }
 
-// EnableTracing controls whether tracing is enabled for this context
-func (s *Span) EnableTracing(enabled bool) {
-	if enabled {
-		s.flags |= tracingFlagEnabled
-	} else {
-		s.flags &= ^tracingFlagEnabled
+// Flags returns flags bitmap. Interpretation of the bits is up to the tracing system.
+func (s Span) Flags() byte { return s.flags }
+
+// SetTraceID sets traceID
+func (s *Span) SetTraceID(traceID uint64) { s.traceID = traceID }
+
+// SetSpanID sets spanID
+func (s *Span) SetSpanID(spanID uint64) { s.spanID = spanID }
+
+// SetParentID sets parentID
+func (s *Span) SetParentID(parentID uint64) { s.parentID = parentID }
+
+// SetFlags sets flags
+func (s *Span) SetFlags(flags byte) { s.flags = flags }
+
+// CurrentSpan extracts OpenTracing Span from the Context, and if found tries to
+// extract zipkin-style trace/span IDs from it using ZipkinSpanFormat carrier.
+// If there is no OpenTracing Span in the Context, an empty span is returned.
+func CurrentSpan(ctx context.Context) *Span {
+	if sp := opentracing.SpanFromContext(ctx); sp != nil {
+		span := &Span{}
+		span.initFromOpenTracing(sp)
+		return span
 	}
+	return &emptySpan
 }
 
-// TracingEnabled checks whether tracing is enabled for this context
-func (s Span) TracingEnabled() bool { return (s.flags & tracingFlagEnabled) == tracingFlagEnabled }
+// emptySpan is returned from CurrentSpan(ctx) when there is no OpenTracing
+// Span in ctx, to avoid returning nil.
+var emptySpan Span
 
-// NewChildSpan begins a new child span in the provided Context
-func (s Span) NewChildSpan() *Span {
-	childSpan := &Span{
-		traceID:  s.traceID,
-		parentID: s.spanID,
-		flags:    s.flags,
+// initFromOpenTracing initializes Span fields from an OpenTracing Span,
+// assuming the tracing implementation supports Zipkin-style span IDs.
+func (s *Span) initFromOpenTracing(span opentracing.Span) {
+	if err := span.Tracer().Inject(span.Context(), ZipkinSpanFormat, s); err != nil {
+		s.initRandom()
```

Comment:
since the span itself isn't stored anywhere, does this mean that multiple calls to `CurrentSpan(ctx)` could return different spans? i think returning the empty span would probably be better in that case